### PR TITLE
beam-search implementation for more exhausting sampling

### PR DESCRIPTION
**Pull request by laurcate**
*Issue #, if available:*

*Description of changes:*
this function largely replaces A2RL Simulator.gpt_sample_n_steps(). It does not
concern states/actions/rewards and only generates the next N tokens using beam search.
This function is to be used by a BYO planner downstream as a replacement of the normal sampling strategy.

Additionally a notebook with example of implementaiton has been added.

*Testing done:*
Yes

## Merge Checklist

*Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're
unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of
what we are going to look for before merging your pull request.*

- [x ] I have read the [CONTRIBUTING](https://github.com/awslabs/amazon-accessible-rl-sdk/blob/main/CONTRIBUTING.md) document.
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I have updated any necessary documentation, including [README](https://github.com/awslabs/amazon-accessible-rl-sdk/blob/main/README.md) and [docs](https://github.com/awslabs/amazon-accessible-rl-sdk/tree/main/docs) (if appropriate).

By submitting this pull request, I confirm that my contribution is made under the terms of the
Apache 2.0 license.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -17,3 +17,5 @@ gym>=0.23.1,<0.26.0
 seaborn
 cloudpickle
 pytorch-lightning>=1.5.0
+
+tensorboardX
```
Comment (verdimrc) [inline requirements.txt:21]:
```suggestion
tensorboardX

```

File: src/a2rl/simulator.py [hunk 2]
```diff
@@ -1248,6 +1248,101 @@ def _handle_unseen_token(self, seq: np.ndarray) -> np.ndarray:
...
+        return_logprobs: bool = False,
+    ):
+        """This function largely replaces A2RL Simulator.gpt_sample_n_steps(). It does not
+        concern states/actions/rewards and only generates the next N tokens using beam search.
+        This function is to be used by a planner.
+
+        Args:
+            seq: A sequence of tokens (1-dimensional only)
+            n_steps: number of tokens to generate
+            beam_width: number of beams used in beam search. Must be <= n of valid tokens in
```
Comment (verdimrc) [inline src/a2rl/simulator.py:1270]: is it `0 < beam_width <= n_steps` or just `beam_width <= n_steps`?

**Review by verdimrc — APPROVED** [review]

**Comment by verdimrc** [comment]
Thank you @laurcate , @patrick22414. Feel free to merge to main.

Ignore the linter error -- looks like a newer Black wants to reformat even files outside this PR.

Recommend you to review the missing coverage in your new function (see the coverage report a few comments above). I'm okay if you want to add extra tests for the edge cases as a separate PR.

**Comment by verdimrc** [comment]
Thank you for the additional test cases.
